### DISCUSSION TITLE
Listing attributes of all nodes in each time-step

### DISCUSSION THREAD

**Discussion by PabArb** [pyswmm/pyswmm#520] [Q&A] [answered]
Hello, I'm pretty new to python and of course to this package. My final objective is to couple a 1D and 2D model for flooding. 
Ideally, I would run the SWMM simulation, extract the flooding for the nodes, create a hotstart file, input the flooding it in the 2D model and run it, update the hotstart with "new flooding values", use_hotstart to resume the simulation. 
For the first approach I'm trying to:

```
with Simulation(r'./MS5_model.inp') as sim:
   step_advance(300)
   allnodes = Nodes(sim)
   for step in sim:
      print(allnodes.flooding)
```

I see in References that it should be iterated through nodes to extract their values, is there a way to do them all at once? I also give a context of my final objective as any recommendations to my approach are well received. 
Thanks, 
Simon

**Comment by bemcdonnell** [comment, marked as answer]
@PabArb, we're certainly here to help!  This sounds like an interesting problem.  So I can be most helpful, are you running into performance limitations with the node iterator to scrape out flooding values or is it a question on "how to"?

The cleanest method to get the flooding values out is using a dictionary comprehension:
```python
node_flooding_dict = {nd.nodeid: nd.flooding for nd in Nodes(sim)}
```
However, if you have a larger model, there could be performance limitations with this approach (maybe the reason you're opening this discussion).  There are some lower level API's inside PySWMM that involve less overhead but if you're newer, they might take a little more time to noodle through. I can help you; feel free to email me directly. 

Regarding your approach, how are you handling the 2D side of the interface?  You made reference to getting data / updating the hotstart file with flooding values.  I'm not sure that approach will work.  If you look at what network values are stored to the hot start file: 

https://github.com/pyswmm/Stormwater-Management-Model/blob/459db1d4dfc61ff994ae01f92eae64e378e08915/src/solver/hotstart.c#L222-L263

`flooding` isn't one of them since it is endogenous to the simulation.  That being said, there is an alternative approach.  If your 2D model is available in Python (or at least callable from Python), PySWMM makes the interfacing very easy.   You do not have to stop/start your simulation or save/use hot start files.  

What you might consider is doing the following:
1. As you are already on track to do this, use the flooding volume to update your surface volume
2. determine the depth of your inundated surface
3. q_inflow = f(d)-losses
4. use PySWMM's `node.generated_inflow(q_inflow)` to introduce that flow back to the hydraulic network (`if node.depth<node.full_depth`)
5. repeat for every time step.  

This code could look like the following.  (For sure this code isn't beautiful or production worthy - just meant to show what is possible.)
 
```python
from pyswmm import Simulation, Nodes

model_eval_timestep = 300 # seconds
current_flooded_vol = 0 # cubic feet

def update_flooded_vol(flow):
    current_flooded_vol += flow * model_eval_timestep
    if current_flooded_vol < 0:
        current_flooded_vol = 0
 
def vol_to_depth_func():
    area = 1000
    depth = current_flooded_vol / area # this equation a function of your topology
    return depth

def inflow_calculation():
   """Inflow function that incorporates losses"""
    q_inflow = vol_to_depth_func() * 0.1
    return q_inflow

with Simulation("fl-path.inp") as sim:
    node_1 = Nodes(sim)['node_xyz']

    sim.step_advance(model_eval_timestep) 
    for step in sim:
        # Get Flooding Rate
        flooding = node_1.flooding
        # Update Surface Volume
        if flooding > 0:
            update_flooded_vol(flooding)
        # Update Flow Exchange
        if node_1.depth < node_1.full_depth # Only if the junction has capacity
            # calculate flow exchange
            q_inflow = inflow_calculation()
            # Push flow into network
            node_1.generated_inflow(q_inflow)
            # Track flow draining back into network
            update_flooded_vol(-1*q_inflow) #<- possible mass balance problem here
        else:
            # Stop flow coming into network
            node_1.generated_inflow(0)
```

The CON to this method is that there is no way to set the HGL to throttle the flooding rate at a node once a node is inundated.  Several researchers have been thinking through those solutions.  However, you might be able to apply a correction factor and push some flow "back" into the network in addition to the q_inflow.  Furthermore, depending on the interfacing step you use, you could have mass balance / continuity concerns that need to be sorted out. 

If you would like additional support, please feel free to email me!

**Comment by PabArb** [comment]
Hello, thanks for your help, I was asking for the "how to". Thanks to your help I've been working on it a bit and have been able to have the flooding values in a list.
Regarding the 2D, I don't really know. I think I will interact with the generate_inflow() in each time step instead of using hotstart files, but I still have to program the 2D overland algorithm. In the meantime, I was wondering if there was a direct way to apply generated_inflow() for all of the nodes in the network according to their nodeid, or if on the contrary I should loop through them. 
Thanks again,
Simon

**Reply by bemcdonnell** [reply]
@PabArb, I’ll be happy to help you worth through an alternative. Ping me by email if you’d like!
